src/api/index.md

```diff
@@ -0,0 +1,277 @@
+# Gutenberg API Prototype
+
+This is proposal for a block API for the new WordPress editor!
+
+Joen Asmussen pointed out that the design of the new editor would be
+"block first". This sunk in and spurred this adventure into creating an API that
+itself is just a single self referencing "block" that can create other blocks
+out of itself.
+
+## Overview
+
+### How does this thing work
+
+While trying to create a block that could make other blocks, which could make
+other blocks, with a nice syntax, I settled on the pattern seen in blockType.js
+et al. Basically, you have a function, that takes a function of properties and
+returns a new reference to a function assigned those properties. This is then
+partially applied to return a new function that takes properties. Each time the
+function is called a new reference is created to enable the ability for new
+entities to be created from a primitive version of itself carrying along its own
+properties and at the beginning starting with none.
+
+### Quick intro
+
+`blockType()`, `block()`, `control()`, and `controlType()`, all serve as an independent
+tool for building larger pieces. Each are identical other than their name. They
+all work from the same fundamental pattern. Let's look at `blockType`, and
+`controlType`. These would serve as a way to create factories for `block()` and
+control(). A good idea might be to have a `create()` method on a type, which would
+create the actual block, or control. Then each block or control would have a
+`render()` method which would create an actual representation of that block or
+control. `render()` could be used to return React components, DOM, vDOM, HTML,
+TinyMCE, whatever is chosen. Let's look at creating a `blockType`.
+
+### Creating a textBlockType
+
+```js
+// textBlock would be the block
+import textBlock from '../blocks/textBlock'
+import blockType from './blockType'
+
+const textBlockTypeProperties = {
+	type: 'wp-text',
+	title: 'Text',
+	controls: [
+		[
+			controlType( {
+				type: 'left-align',
+				title: 'Align Left',
+				displayArea: 'block'
```

Comment:
> Do you see an advantage to making it a property of the control itself?
Yes.

This ties in with the above, comment.  If what I am saying doesn't make sense I can Google hangout or something and probably explain it better.

Think of everything as data.  It can be transformed and turned into any format you want.  If you have these as properties on the control object itself, then when iterating/looping (whatever you want to call it ) you can create new lists of controls more easily.  If you prematurely group the controls as an object to get all of the controls to iterate over you would need to iterate over `Object.keys()` first and then match that against the object to get the actual value of the control.  This gets really cumbersome when you are doing deep iteration and when you want to iterate across all of the controls on a different grouping.  Keeping controls as an array then using `map()` or `forEach()` or a regular old for loop, for me makes it easier to work with them; just a style choice I guess.  But typically if you keep them as an array of objects it will make them more flexible, and lead to less code writing.